File: codigo/Back/models/demanda.py
from enum import Enum

from sqlalchemy import Column, Date, Enum as SqlEnum, Float, Integer, String

from database.database import Base


class DemandaStatus(str, Enum):
    PENDENTE = "PENDENTE"
    ACEITO = "ACEITO"
    EM_EXECUCAO = "EM_EXECUCAO"
    CONCLUIDO = "CONCLUIDO"


class Demanda(Base):
    __tablename__ = "demandas"

    id = Column(Integer, primary_key=True, index=True)
    cliente_id = Column(Integer, nullable=False, index=True)
    prestador_id = Column(Integer, nullable=True, index=True)
    titulo = Column(String, nullable=False)
    descricao = Column(String, nullable=False)
    origem = Column(String, nullable=False)
    destino = Column(String, nullable=True)
    area_hectares = Column(Float, nullable=False)
    valor_recompensa = Column(Float, nullable=False)
    tipo_servico = Column(String, nullable=False)
    data_limite = Column(Date, nullable=True)
    status = Column(
        SqlEnum(DemandaStatus, name="demanda_status"),
        default=DemandaStatus.PENDENTE,
        nullable=False,
    )

File: codigo/Mobile/field_flow/lib/models/demanda.dart
import 'enums.dart';

/// Solicitacao de servico criada por um cliente (DemandaResponse no backend).
class Demanda {
  Demanda({
    required this.id,
    required this.clienteId,
    required this.titulo,
    required this.descricao,
    required this.origem,
    required this.areaHectares,
    required this.unidadePagamento,
    required this.tipoServico,
    required this.status,
    this.destino,
    this.valorRecompensa,
    this.dataLimite,
    this.prestadorId,
  });

  final int id;
  final int clienteId;
  final String titulo;
  final String descricao;
  final String origem;
  final double areaHectares;
  final UnidadePagamento unidadePagamento;
  final String tipoServico;
  final DemandaStatus status;
  final String? destino;
  final double? valorRecompensa;
  final DateTime? dataLimite;
  final int? prestadorId;

  factory Demanda.fromJson(Map<String, dynamic> json) => Demanda(
        id: json['id'] as int,
        clienteId: json['cliente_id'] as int,
        titulo: json['titulo'] as String,
        descricao: json['descricao'] as String,
        origem: json['origem'] as String,
        areaHectares: (json['area_hectares'] as num).toDouble(),
        unidadePagamento:
            UnidadePagamento.fromWire(json['unidade_pagamento'] as String),
        tipoServico: json['tipo_servico'] as String,
        status: DemandaStatus.fromWire(json['status'] as String),
        destino: json['destino'] as String?,
        valorRecompensa: (json['valor_recompensa'] as num?)?.toDouble(),
        dataLimite: json['data_limite'] != null
            ? DateTime.tryParse(json['data_limite'] as String)
            : null,
        prestadorId: json['prestador_id'] as int?,
      );
}
